overlord/snapstate/snapstate.go

```diff
@@ -529,13 +536,41 @@ func Update(st *state.State, name, channel string, revision snap.Revision, userI
 	return doInstall(st, &snapst, snapsup)
 }
 
+func validInfoForFlags(info *snap.Info, snapst *SnapState, flags Flags) bool {
+	switch c := info.Confinement; c {
+	case snap.StrictConfinement, "":
+		// strict is always fine
+		return true
+	case snap.DevModeConfinement:
+		// --devmode needs to be specified every time (==> ignore snapst)
+		return flags.DevModeAllowed()
+	case snap.ClassicConfinement:
+		if flags.Classic {
+			return true
+		}
+
+		if snapst != nil && snapst.Flags.Classic {
+			return true
+		}
+
+		return false
+	default:
+		logger.Panicf("unknown confinement %q", c)
```

Comment:
Seems better to report an error here instead of panicking, as this is data coming from outside. Otherwise we'll need to make sure that all entry point necessarily validate the potential confinements upfront.